src/haz3lcore/dynamics/TypeAssignment.re

```diff
@@ -0,0 +1,293 @@
+open Util;
+open OptUtil.Syntax;
+
+let equal_typ_list = (l: list(Typ.t)): option(Typ.t) => {
+  switch (l) {
+  | [] => None
+  | [ty, ..._] =>
+    List.fold_left((acc, t) => {acc && Typ.eq(t, ty)}, true, l)
+      ? Some(ty) : None
+  };
+};
+
+let arrow_aux = (ty: Typ.t): Typ.t => {
+  switch (ty) {
+  | Unknown(Internal) => Arrow(Unknown(Internal), Unknown(Internal))
+  | _ => ty
+  };
+};
+
+let delta_ty = (id: MetaVar.t, m: Statics.Map.t): option(Typ.t) => {
+  switch (Id.Map.find_opt(id, m)) {
+  | Some(InfoExp({mode, ctx, _})) =>
+    switch (mode) {
+    | Syn
+    | SynFun => Some(Unknown(Internal))
+    | Ana(ana_ty) => Some(Typ.normalize(ctx, ana_ty))
+    }
+  | _ => None
+  };
+};
+
+let ground = (ty: Typ.t): bool => {
```

Comment:
there should be more ground types than this -- ground products and sums, and I think there is already a function to determine this in the evaluator?